opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConcurrentConfigurationControllerTest.java

```diff
@@ -0,0 +1,188 @@
+        MockitoAnnotations.initMocks(this);
+        return new ResourceConfig(ConfigurationController.class)
+                .register(new AbstractBinder() {
+                    @Override
+                    protected void configure() {
+                        bind(suggesterService).to(SuggesterService.class);
+                    }
+                });
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        origSourceRootPath = env.getSourceRootPath();
+        origDataRootPath = env.getDataRootPath();
+        origProjects = env.getProjects();
+        origRepositories = env.getRepositories();
+
+        // prepare test repository
+        repository = new TestRepository();
+        repository.create(HistoryGuru.class.getResourceAsStream("repositories.zip"));
+
+        env.setSourceRoot(repository.getSourceRoot());
+        env.setDataRoot(repository.getDataRoot());
+
+
+        List<RepositoryInfo> repositoryInfos = new ArrayList<>();
+        Map<String, Project> projects = new TreeMap<>();
+
+        /*
+         * Prepare PROJECTS_COUNT git repositories, named project-{i} in the test repositories directory.
+         */
+        for (int i = 0; i < PROJECTS_COUNT; i++) {
+            Project project = new Project();
+            project.setName("project-" + i);
+            project.setPath("/project-" + i);
+            RepositoryInfo repo = new RepositoryInfo();
+            repo.setDirectoryNameRelative("/project-" + i);
+
+            projects.put("project-" + i, project);
+            repositoryInfos.add(repo);
+
+            // create the repository
+            FileUtils.copyDirectory(
+                    Paths.get(repository.getSourceRoot(), "git").toFile(),
+                    Paths.get(repository.getSourceRoot(), "project-" + i).toFile()
+            );
+        }
+
+        env.setRepositories(repositoryInfos);
+        env.setProjects(projects);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        repository.destroy();
+
+        env.setProjects(origProjects);
+        env.setRepositories(origRepositories);
+        env.setSourceRoot(origSourceRootPath);
+        env.setDataRoot(origDataRootPath);
+    }
+
+    @Test
+    @ConditionalRun(RepositoryInstalled.GitInstalled.class)
+    public void testConcurrentConfigurationReloads() throws InterruptedException, ExecutionException {
+        final ExecutorService threadPool = Executors.newFixedThreadPool(THREAD_COUNT);
+        List<Future<?>> futures = new LinkedList<>();
+
+        /*
+         * Now run setting a value in parallel which triggers configuration reload.
+         */
+        for (int i = 0; i < TASK_COUNT; i++) {
+            futures.add(threadPool.submit(() -> {
+                Response put = target("configuration")
+                        .path("projectsEnabled")
+                        .request()
+                        .put(Entity.text("true"));
```

Comment:
maybe it should set the true/false property value randomly - in case someone implements smarter property setting (that would become no-op if the value is already set)

Reply:
then there should be a different test, this one would become needleslly complex

Reply:
the assertions do not care about the property value.

Reply:
perhaps use something like:
```java
Random rand = new Random(System.currentTimeMillis());
for (...) {
    ...put(rand.nextInt(2) == 1 ? Entity.text("true") : Entity.text("false")
}
```

Reply:
Assertions do care, if you set enabledProjects = false, then you can't expect 20 projects in the next assertion.

Reply:
I see. Let's leave it then.